mantis/simulation/routers/lautaro.py

```diff
@@ -0,0 +1,201 @@
+# solves using NLP optimization (or what best underlying engine decides)
+# Models cross chain transfers as fees as """pools"""
+# Uses decision variables to decide if to do Transfer to tap pool or not.
+
+import numpy as np
+import cvxpy as cp
+import copy 
+import threading as th
+import os
+MAX_RESERVE = 1e10
+
+from simulation.routers.data import AllData, Input, TId, TNetworkId, Ctx, AssetTransfers, AssetPairsXyk, TAmount
+
+class Edge:
+    U : list[int]
+    B : list[TAmount]
+    W : list[int]
+    F : list[float]
+
+    def toFloat(self, x):
+        try:
+            return float(x)
+        except:
+            return 0.0
+
+    def __init__(self, e : [AssetTransfers|AssetPairsXyk], tokensIds : dict[TId, int]): 
+        if isinstance(e,AssetTransfers):
+            self.__initFromTransfers(e, tokensIds)
+        else:
+            self.__initFromPairsXyk(e, tokensIds)
+
+    def __initFromTransfers(self, e : AssetTransfers, tokensIds : dict[TId, int]):
+        self.U = [tokensIds[e.in_asset_id], tokensIds[e.out_asset_id]]
+        self.B = [e.amount_of_in_token, e.amount_of_out_token]
+        self.W = [1,1]
+        self.F = [float(e.fee_per_million)/1_000_000.0, float(e.fee_per_million)/1_000_000.0]
+    
+    def __initFromPairsXyk(self, e : AssetPairsXyk, tokensIds : dict[TId, int]):
+        self.U = [tokensIds[e.in_asset_id], tokensIds[e.out_asset_id]]
+        self.B = [e.in_token_amount, e.out_token_amount]
+        self.W = [e.weight_of_a, e.weight_of_b]
+        self.F = [self.toFloat(e.fee_in), self.toFloat(e.fee_out)]
+
+    def GetAmount(self, Ti, Xi):
+        i,o = 0,1
+        if Ti == self.U[1]:
+            i,o = 1,0
+        Xi = Xi * (1-self.F[i])
+        return self.B[o] * (1-(self.B[i]/(self.B[i]+Xi))**(self.W[i]/self.W[o]))
+
+    def DoChange(self, Ti, Xi):
+        i,o = 0,1
+        if Ti == self.U[1]:
+            i,o = 1,0
+        Xi = Xi * (1-self.F[i])
+        result = Xi * (self.B[i]/(self.B[i]+Xi))**(self.W[i]/self.W[o])
+        self.B[i] += Xi
+        self.B[o] -= result        
+        return result
+    
+    def GetOther(self, Ti):
+        if Ti == self.U[0]:
+            return self.U[1]
+        return self.U[0]
+
```

Comment:
Estado?

Reply:
Estado is the spanish word for State, it's a class to pass the common state to the paralell functions.